src/DuneTaskProvider.ml

```diff
@@ -0,0 +1,108 @@
+open Import
+
+type t = Disposable.t
+
+let dune_bin = "dune"
+
+let task_type = "dune"
+
+let taskDefinition = { Task.type_ = task_type }
+
+let source = task_type
+
+let problemMatchers = [| "$ocamlc" |]
+
+(* the ocamlc matcher is not able to parse ocaml compiler errors unless they
+   follow the short style. *)
+let env = Js.Dict.fromList [ ("OCAML_ERROR_STYLE", "short") ]
+
+module Setting = struct
+  type t = bool
+
+  let ofJson json =
+    let open Json.Decode in
+    bool json
+
+  let toJson (t : t) =
+    let open Json.Encode in
+    bool t
+
+  let t =
+    Settings.create ~scope:Workspace ~key:"dune.autoDetect" ~ofJson ~toJson
+end
+
+let folderRelativePath folders file =
+  Array.fold_left
+    (fun acc (folder : Folder.t) ->
+      match acc with
+      | Some _ -> acc
+      | None -> (
+        match String.chopPrefix file ~prefix:folder.uri.fsPath with
+        | None -> acc
+        | Some withoutPrefix -> Some (folder, withoutPrefix) ))
+    None folders
+
+let commandLine () =
+  let open Promise.O in
+  Toolchain.ofSettings () >>| function
+  | None -> sprintf "%s build" dune_bin
+  | Some pm ->
+    let resources = Toolchain.makeResources pm in
+    let cmd, args = Toolchain.getCommand resources dune_bin [ "build" ] in
+    Js.Array.joinWith " " (Js.Array.concat args [| Path.toString cmd |])
+
+let computeTasks cancellationToken =
+  let open Promise.O in
+  let folders = Workspace.workspaceFolders () in
+  let excl =
+    (* ignoring dune files from _build, _opam, _esy *)
+    Some "{**/_*}"
+  in
+  let inc = "**/{dune,dune-project,dune-workspace}" in
```

Comment:
this could be built based on a configuration setting if you want to have a way to only pick the dune-project/workspace files